core/src/components/searchbar/searchbar.tsx

```diff
@@ -240,45 +250,86 @@ export class Searchbar implements ComponentInterface {
     return Promise.resolve(this.nativeInput!);
   }
 
+  /**
+   * Emits an `ionChange` event.
+   *
+   * This API should be called for user committed changes.
+   * This API should not be used for external value changes.
+   */
+  private emitValueChange() {
+    const { value } = this;
+    // Checks for both null and undefined values
+    const newValue = value == null ? value : value.toString();
+    // Emitting a value change should update the internal state for tracking the focused value
+    this.focusedValue = newValue;
+    this.ionChange.emit({ value: newValue });
+  }
+
   /**
    * Clears the input field and triggers the control change.
    */
-  private onClearInput = (shouldFocus?: boolean) => {
+  private onClearInput = async (shouldFocus?: boolean) => {
     this.ionClear.emit();
 
-    // setTimeout() fixes https://github.com/ionic-team/ionic/issues/7527
-    // wait for 4 frames
-    setTimeout(() => {
-      const value = this.getValue();
-      if (value !== '') {
-        this.value = '';
-        this.ionInput.emit();
-
-        /**
-         * When tapping clear button
-         * ensure input is focused after
-         * clearing input so users
-         * can quickly start typing.
-         */
-        if (shouldFocus && !this.focused) {
-          this.setFocus();
+    return new Promise((resolve) => {
+      // setTimeout() fixes https://github.com/ionic-team/ionic/issues/7527
+      // wait for 4 frames
+      setTimeout(() => {
+        const value = this.getValue();
+        if (value !== '') {
+          this.value = '';
+          this.ionInput.emit();
+
+          /**
+           * When tapping clear button
+           * ensure input is focused after
+           * clearing input so users
+           * can quickly start typing.
+           */
+          if (shouldFocus && !this.focused) {
+            this.setFocus();
+
+            /**
+             * The setFocus call above will clear focusedValue,
+             * but ionChange will never have gotten a chance to
+             * fire. Manually revert focusedValue so onBlur can
+             * compare against what was in the box before the clear.
+             */
+            this.focusedValue = value;
+          }
         }
-      }
-    }, 16 * 4);
+
+        resolve('cleared');
+      }, 16 * 4);
+    });
   };
 
   /**
    * Clears the input field and tells the input to blur since
    * the clearInput function doesn't want the input to blur
    * then calls the custom cancel function if the user passed one in.
    */
-  private onCancelSearchbar = (ev?: Event) => {
+  private onCancelSearchbar = async (ev?: Event) => {
     if (ev) {
       ev.preventDefault();
       ev.stopPropagation();
     }
     this.ionCancel.emit();
-    this.onClearInput();
+
+    // get cached values before clearing the input
+    const value = this.getValue();
+    const focused = this.focused;
+
+    await this.onClearInput();
```

Comment:
I'm noticing that the `ionChange` event fires before the `ionInput` event here.

Reply:
Admittedly I can't figure out why 🤔 The `ionInput.emit` call is definitely made before `ionChange.emit`, but the actual event doesn't get picked up by listeners until after. Maybe something with how Stencil is emitting the event(s)? Do you think this could cause problems?

Reply:
The problem is that the debounce function is not being removed when `debounce="0"`. So in this case, the `ionInput` event is being emitted inside of `setTimeout(() => { ... }, 0);`. This will cause it to be emitted after the `ionChange` event, which gets emitted synchronously.

I'd argue that this is a bug in how `debounce` was originally implemented. Due to how `ionChange` used to function, the team probably never considered this case.

Reply:
Discussed offline: We are going to have `debounce` default to `undefined` and remove the debounce effect in that case. A debounce of `0` will remain as a `setTimeout`.